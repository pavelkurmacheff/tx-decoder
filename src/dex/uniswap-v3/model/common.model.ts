import {BigNumber} from '@ethersproject/bignumber';
import {Interface} from '@ethersproject/abi';
import { CustomTokensService } from '../../../helpers/tokens/custom-tokens.service';
import { Token } from 'src/core/transaction-rich/token';



export interface BlockchainResources {
    readonly tokens: {[key: string]: Token}, // https://tokens.1inch.io/v1.1/1
    readonly tokenPrices: {[key: string]: string}, // https://token-prices.1inch.io/v1.1/1
}

export interface Web3Resources extends BlockchainResources {
    readonly customTokens: CustomTokensService,
}

export interface BlockchainRpcCaller {
    rpcUrl: string;

    call<T>(method: string, params: unknown[]): Promise<T>;
}

export interface DecodeInfo {
    iface: Interface;
    methodSelector: string;
}
