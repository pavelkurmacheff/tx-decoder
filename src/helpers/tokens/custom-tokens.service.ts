import {
    ChainTokenByNetwork,
    NATIVE_TOKEN_ADDRESS,
    ZERO_ADDRESS,
} from '../../core/const/common.const';
import {Web3Service} from '../web3/web3.service';
import {ERC20, ERC20_BYTE32, ERC20_UPPER_CASE} from './abi';
import {AbiItem} from 'web3-utils';
import {Token} from '../../core/token';
import {ChainId} from '../../core/chain-id';

export class CustomTokensService {
    constructor(
        private cache: Map<String, Token>,
        private web3Service: Web3Service, 
        readonly chainId: ChainId) {

        const t: Token = ChainTokenByNetwork[this.chainId]
            ? ChainTokenByNetwork[this.chainId]
            : ChainTokenByNetwork[ChainId.Ethereum];

        this.cache.set(ZERO_ADDRESS, t);
        this.cache.set(NATIVE_TOKEN_ADDRESS, t);
    }

    async getTokenByAddress(address: string): Promise<Token | null> {
        try {
            const cachedArrd = this.cache.get(address);
            if (cachedArrd) {
                return cachedArrd;
            }

            const data = await this.fetchTokenInfo(address.toLowerCase());
            if (!data || !data.decimals) {
                return null;
            }

            const info = this.buildMinimizedTokenFromData(data);
            this.cache.set(address, info);

            return info;
        } catch (e) {
            console.error(e);
            return null;
        }
    }

    
    private buildMinimizedTokenFromData(data: Token): Token {
        const info: Token = {
            name: data.name,
            symbol: data.symbol,
            address: data.address,
            decimals: data.decimals,
        };

        if (data.logoURI) {
            info.logoURI = data.logoURI;
        }
        return info;
    }

    private async fetchTokenInfo(address: string): Promise<Token | null> {
        try {
            // eslint-disable-next-line @typescript-eslint/no-misused-promises,no-async-promise-executor
            return new Promise(async (resolve, reject) => {
                try {
                    resolve(await this.fetchNormalTokenInfo(address));
                } catch (e) {
                    if (isUpperCaseToken(e)) {
                        try {
                            resolve(await this.fetchUpperCaseTokenInfo(address));
                        } catch (e) {
                            console.error(e);
                        }
                    } else if (isBytes32TokenInfo(e)) {
                        try {
                            resolve(await this.fetchBytes32TokenInfo(address));
                        } catch (e) {
                            console.error(e);
                        }
                    }
                }
                reject('cannot fetch details for token: ' + address);
            });
        } catch (e) {
            return null;
        }
    }

    private async fetchNormalTokenInfo(tokenAddress: string): Promise<Token> {
        const token = new this.web3Service.web3.eth.Contract(
            ERC20 as AbiItem[] | AbiItem,
            tokenAddress
        );

        return {
            symbol: await token.methods.symbol().call(),
            name: await token.methods.name().call(),
            address: tokenAddress,
            decimals: Number(
                (await token.methods.decimals().call()).toString()
            ),
        };
    }

    private async fetchUpperCaseTokenInfo(tokenAddress: string): Promise<Token> {
        const token = new this.web3Service.web3.eth.Contract(
            ERC20_UPPER_CASE as AbiItem[] | AbiItem,
            tokenAddress
        );

        return {
            symbol: await token.methods.SYMBOL().call(),
            name: await token.methods.NAME().call(),
            address: tokenAddress,
            decimals: Number(
                (await token.methods.DECIMALS().call()).toString()
            ),
        };
    }

    private async fetchBytes32TokenInfo(tokenAddress: string): Promise<Token> {
        const token = new this.web3Service.web3.eth.Contract(
            ERC20_BYTE32 as AbiItem[] | AbiItem,
            tokenAddress
        );

        return {
            symbol: this.web3Service.web3.utils.hexToUtf8(
                await token.methods.symbol().call()
            ),
            name: this.web3Service.web3.utils.hexToUtf8(
                await token.methods.name().call()
            ),
            address: tokenAddress,
            decimals: Number(
                (await token.methods.decimals().call()).toString()
            ),
        };
    }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function isUpperCaseToken(e: any): boolean {
    return e.toString().indexOf('Reverted 0x') !== -1;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function isBytes32TokenInfo(e: any): boolean {
    return (
        e.toString().indexOf('Number can only safely store up to 53 bits') !==
            -1 || e.toString().indexOf('overflow') !== -1
    );
}

export const isETH = (address: string): boolean => {
    return (
        NATIVE_TOKEN_ADDRESS === address.toLowerCase() ||
        ZERO_ADDRESS === address
    );
};
