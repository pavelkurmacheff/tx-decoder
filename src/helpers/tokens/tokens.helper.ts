import {
    BlockchainResources,
    Web3Resources,
} from '../../dex/uniswap-v3/model/common.model';
import { NATIVE_TOKEN_ADDRESS, ZERO_ADDRESS } from '../../core/const/common.const';
import { Token } from 'src/core/token';

export function findTokenByAddress(resources: BlockchainResources, tokenAddress: string): Token | null {
    const address =
        tokenAddress === ZERO_ADDRESS
            ? NATIVE_TOKEN_ADDRESS
            : tokenAddress.toLowerCase();

    return resources.tokens[address] || null;
}

export async function tryToFindTokenByAddress(resources: Web3Resources, tokenAddress: string): Promise<Token | null> {
    let token = findTokenByAddress(resources, tokenAddress);
    if (!token) {
        token = await resources.customTokens.getTokenByAddress(tokenAddress);
    }
    return token || null;
}
