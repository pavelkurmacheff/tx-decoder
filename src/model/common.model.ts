export interface Transaction {
    from: string;
    to: string;
    value: string;
    data: string;
    gasPrice: string | number;
    gasLimit: string | number;
    nonce?: string | number;
}

export interface Token {
    address: string;
    name: string;
    decimals: number;
    symbol: string;
}

export interface BlockchainResources {
    readonly tokens: {[key: string]: Token}, // https://tokens.1inch.io/v1.1/1
    readonly tokenPrices: {[key: string]: string}, // https://token-prices.1inch.io/v1.1/1
}
