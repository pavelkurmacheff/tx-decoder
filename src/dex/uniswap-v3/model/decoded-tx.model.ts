import {SwapTxDecoded} from './swap-tx.model';
import {MultipleTxsDecoded} from './multiple-tx.model';
import {UnwrapTxDecoded} from './unwrap-tx.model';
import {ApproveTxDecoded} from '../../../core/transaction-rich/approve-tx.model';
import {LimitOrderFillDecoded} from '../../../core/transaction-rich/limit-order-fill.model';

export type DecodedTx =
    | ApproveTxDecoded
    | SwapTxDecoded
    | UnwrapTxDecoded
    | MultipleTxsDecoded
    | LimitOrderFillDecoded;
