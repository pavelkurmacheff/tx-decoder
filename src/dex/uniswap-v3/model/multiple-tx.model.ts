import { SwapTxDecoded } from './swap-tx.model';
import { UnwrapTxDecoded } from './unwrap-tx.model';
import { ApproveTxDecoded } from '../../../core/transaction-rich/approve-tx.model';

export interface MultipleTxsDecoded {
    readonly txs: (SwapTxDecoded | UnwrapTxDecoded | ApproveTxDecoded)[];
    readonly error?: Error;
}
