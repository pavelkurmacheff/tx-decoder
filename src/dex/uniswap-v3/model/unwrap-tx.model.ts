import { Token } from './common.model';
import { BigNumber } from '@ethersproject/bignumber';

export interface UnwrapTxDecoded {
    readonly token: Token;
    readonly amount?: BigNumber;
    readonly minReturnAmount?: BigNumber;
    readonly error?: Error;
}

