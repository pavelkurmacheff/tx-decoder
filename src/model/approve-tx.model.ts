import { Token } from './common.model';
import { BigNumber } from '@ethersproject/bignumber';

export interface ApproveTxDecoded {
    token: Token;
    value: BigNumber;
    time?: number;
}
