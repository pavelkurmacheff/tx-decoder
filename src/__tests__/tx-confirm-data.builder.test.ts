import {Transaction} from '../model/common.model';
import {TxConfirmDataBuilder} from '../tx-confirm-data.builder';
import {TOKENS_MOCK} from './mocks/tokens.mock';
import {TOKEN_PRICES_MOCK} from './mocks/token-prices.mock';

describe('Builder of transaction data for confirmation', () => {
    let txUiItemsBuilder: TxConfirmDataBuilder;

    beforeEach(() => {
        txUiItemsBuilder = new TxConfirmDataBuilder({
            tokens: TOKENS_MOCK,
            tokenPrices: TOKEN_PRICES_MOCK
        });
    });

    // https://etherscan.io/tx/0xbd44b75afac45ef81af448c47b40d32ac705fe2d0354b39097f5ebfd62046ffe
    it('Approve transaction', () => {
        const txConfig: Transaction = {
            nonce: 383,
            gasPrice: '0x1da4f97c6e',
            gasLimit: '0x011150',
            from: '0x3b608c5243732903152e38f1dab1056a4a79b980',
            to: '0x4fabb145d64652a948d72533023f6e7a623c7c53',
            value: '0x00',
            data: '0x095ea7b30000000000000000000000001111111254fb6c44bac0bed2854e76f90643097dffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff'
        };

        const {items, txType} = txUiItemsBuilder.buildItemsForTx(txConfig);

        expect(txType).toBe('approve');
        expect(items).toMatchSnapshot();
    });

    // https://etherscan.io/tx/0x698156119b255c33347bf844a0245dafdc39e871f3825ae5c24cc0019be748ed
    it('Swap transaction: swap()', () => {
        const txConfig: Transaction = {
            nonce: 194,
            gasPrice: '0x1b7b430f2c',
            gasLimit: '0x091d06',
            from: '0x64741d0b9e376d75873c12e1b0cdccc26c3bcb04',
            to: '0x1111111254fb6c44bac0bed2854e76f90643097d',
            value: '0x00',
            data: '0x7c025200000000000000000000000000220bda5c8994804ac96ebe4df184d25e5c2196d400000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000180000000000000000000000000dac17f958d2ee523a2206206994597c13d831ec7000000000000000000000000853d955acef822db058eb8505911ed77f175b99e000000000000000000000000220bda5c8994804ac96ebe4df184d25e5c2196d400000000000000000000000064741d0b9e376d75873c12e1b0cdccc26c3bcb040000000000000000000000000000000000000000000000000000001aa580321900000000000000000000000000000000000000000000181b5805bbc9ae4bb3e600000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000004e0000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000030000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000016000000000000000000000000000000000000000000000000000000000000002808000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000064eb5625d9000000000000000000000000dac17f958d2ee523a2206206994597c13d831ec7000000000000000000000000d632f22692fac7611d2aa1c0d552930d43caed3b0000000000000000000000000000000000000000000000000000001aa580321900000000000000000000000000000000000000000000000000000000800000000000000000000000d632f22692fac7611d2aa1c0d552930d43caed3b000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000084a6417ed6000000000000000000000000000000000000000000000000000000000000000300000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001aa5803219000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000018414284aab00000000000000000000000000000000000000000000000000000000000000808000000000000000000000000000000000000000000000000000000000000044000000000000000000000000853d955acef822db058eb8505911ed77f175b99e00000000000000000000000000000001000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000064d1660f99000000000000000000000000853d955acef822db058eb8505911ed77f175b99e0000000000000000000000001111111254fb6c44bac0bed2854e76f90643097d00000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e26b9977'
        };

        const {items, txType} = txUiItemsBuilder.buildItemsForTx(txConfig);

        expect(txType).toBe('swap');
        expect(items).toMatchSnapshot();
    });
});
