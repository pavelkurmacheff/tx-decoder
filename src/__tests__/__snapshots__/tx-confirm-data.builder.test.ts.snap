// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Builder of transaction data for confirmation Approve transaction 1`] = `
Array [
  Object {
    "key": Object {
      "type": "placeholder",
    },
    "value": Object {
      "type": "localizable",
      "value": Object {
        "args": Array [
          "BUSD",
        ],
        "format": "To continue, you need to allow 1inch smart contracts to use your %%0%%. This has to be done only once for each token.",
      },
    },
  },
  Object {
    "key": Object {
      "type": "token",
      "value": Object {
        "token": Object {
          "address": "0x4fabb145d64652a948d72533023f6e7a623c7c53",
          "decimals": 18,
          "logoURI": "https://tokens.1inch.io/0x4fabb145d64652a948d72533023f6e7a623c7c53.png",
          "name": "Binance USD",
          "symbol": "BUSD",
        },
        "type": "amount",
      },
    },
    "value": Object {
      "type": "amount",
      "value": Object {
        "sign": "0",
        "token": Object {
          "address": "0x4fabb145d64652a948d72533023f6e7a623c7c53",
          "decimals": 18,
          "logoURI": "https://tokens.1inch.io/0x4fabb145d64652a948d72533023f6e7a623c7c53.png",
          "name": "Binance USD",
          "symbol": "BUSD",
        },
        "value": "115792089237316195423570985008687907853269984665640564039457584007913129639935",
      },
    },
  },
  Object {
    "key": Object {
      "type": "wallet",
      "value": Object {
        "type": "own",
      },
    },
    "value": Object {
      "type": "text",
      "value": Object {
        "text": "0x3b608c5243732903152e38f1dab1056a4a79b980",
        "type": "address",
      },
    },
  },
  Object {
    "key": Object {
      "type": "placeholder",
    },
    "value": Object {
      "type": "placeholder",
      "value": Object {
        "type": "fee",
      },
    },
  },
]
`;

exports[`Builder of transaction data for confirmation Swap transaction: swap() 1`] = `
Array [
  Object {
    "key": Object {
      "type": "token",
      "value": Object {
        "token": Object {
          "address": "0xdac17f958d2ee523a2206206994597c13d831ec7",
          "decimals": 6,
          "logoURI": "https://tokens.1inch.io/0xdac17f958d2ee523a2206206994597c13d831ec7.png",
          "name": "Tether USD",
          "symbol": "USDT",
        },
        "type": "from",
      },
    },
    "value": Object {
      "type": "amount",
      "value": Object {
        "sign": "-",
        "token": Object {
          "address": "0xdac17f958d2ee523a2206206994597c13d831ec7",
          "decimals": 6,
          "logoURI": "https://tokens.1inch.io/0xdac17f958d2ee523a2206206994597c13d831ec7.png",
          "name": "Tether USD",
          "symbol": "USDT",
        },
        "value": Object {
          "hex": "0x1aa5803219",
          "type": "BigNumber",
        },
      },
    },
  },
  Object {
    "key": Object {
      "type": "token",
      "value": Object {
        "token": Object {
          "address": "0x853d955acef822db058eb8505911ed77f175b99e",
          "decimals": 18,
          "logoURI": "https://tokens.1inch.io/0x853d955acef822db058eb8505911ed77f175b99e.png",
          "name": "Frax",
          "symbol": "FRAX",
        },
        "type": "to",
      },
    },
    "value": Object {
      "type": "amount",
      "value": Object {
        "sign": "+",
        "token": Object {
          "address": "0x853d955acef822db058eb8505911ed77f175b99e",
          "decimals": 18,
          "logoURI": "https://tokens.1inch.io/0x853d955acef822db058eb8505911ed77f175b99e.png",
          "name": "Frax",
          "symbol": "FRAX",
        },
        "value": "113841200360986751251430",
      },
    },
  },
  Object {
    "key": Object {
      "type": "wallet",
      "value": Object {
        "type": "own",
      },
    },
    "value": Object {
      "type": "text",
      "value": Object {
        "text": "0x64741d0b9e376d75873c12e1b0cdccc26c3bcb04",
        "type": "address",
      },
    },
  },
  Object {
    "key": Object {
      "type": "placeholder",
    },
    "value": Object {
      "type": "placeholder",
      "value": Object {
        "type": "fee",
      },
    },
  },
  Object {
    "key": Object {
      "type": "title",
      "value": Object {
        "title": "Sell price",
      },
    },
    "value": Object {
      "type": "rate",
      "value": Object {
        "destinationAmount": "1000000000000000000",
        "destinationToken": Object {
          "address": "0x853d955acef822db058eb8505911ed77f175b99e",
          "decimals": 18,
          "logoURI": "https://tokens.1inch.io/0x853d955acef822db058eb8505911ed77f175b99e.png",
          "name": "Frax",
          "symbol": "FRAX",
        },
        "sourceAmount": "1000000",
        "sourceToken": Object {
          "address": "0xdac17f958d2ee523a2206206994597c13d831ec7",
          "decimals": 6,
          "logoURI": "https://tokens.1inch.io/0xdac17f958d2ee523a2206206994597c13d831ec7.png",
          "name": "Tether USD",
          "symbol": "USDT",
        },
      },
    },
  },
  Object {
    "key": Object {
      "type": "title",
      "value": Object {
        "title": "Buy price",
      },
    },
    "value": Object {
      "type": "rate",
      "value": Object {
        "destinationAmount": "1000000",
        "destinationToken": Object {
          "address": "0xdac17f958d2ee523a2206206994597c13d831ec7",
          "decimals": 6,
          "logoURI": "https://tokens.1inch.io/0xdac17f958d2ee523a2206206994597c13d831ec7.png",
          "name": "Tether USD",
          "symbol": "USDT",
        },
        "sourceAmount": "1000000000000000000",
        "sourceToken": Object {
          "address": "0x853d955acef822db058eb8505911ed77f175b99e",
          "decimals": 18,
          "logoURI": "https://tokens.1inch.io/0x853d955acef822db058eb8505911ed77f175b99e.png",
          "name": "Frax",
          "symbol": "FRAX",
        },
      },
    },
  },
  Object {
    "key": Object {
      "type": "title",
      "value": Object {
        "title": "Minimum Received",
      },
    },
    "value": Object {
      "type": "amount",
      "value": Object {
        "sign": "0",
        "token": Object {
          "address": "0x853d955acef822db058eb8505911ed77f175b99e",
          "decimals": 18,
          "logoURI": "https://tokens.1inch.io/0x853d955acef822db058eb8505911ed77f175b99e.png",
          "name": "Frax",
          "symbol": "FRAX",
        },
        "value": Object {
          "hex": "0x181b5805bbc9ae4bb3e6",
          "type": "BigNumber",
        },
      },
    },
  },
]
`;
